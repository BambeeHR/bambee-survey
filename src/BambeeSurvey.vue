<template>
  <SurveyComponent :survey="model" class="base-survey" />
</template>

<script>
import * as SurveyVue from 'survey-vue';
import cloneDeep from 'lodash.clonedeep';
import resolveConfig from 'tailwindcss/resolveConfig';
import tailwindConfig from '../tailwind.config';
import 'survey-vue/survey.css';

const { theme } = resolveConfig(tailwindConfig);

const SurveyComponent = SurveyVue.Survey;

export default {
  components: { SurveyComponent },
  props: {
    survey: {
      type: Object,
      required: true,
    },
    initialValues: {
      type: Object,
      default: () => ({}),
    },
    mode: {
      type: String,
      default: 'edit',
    },
  },
  data() {
    const json = cloneDeep(this.survey);
    const model = new SurveyVue.Model(json);
    model.mergeData(this.initialValues);

    return { model };
  },
  created() {
    this.model.onValueChanged.add((sender, options) => {
      this.$emit('input', sender, options);
    });

    this.model.onComplete.add((sender) => {
      this.$emit('complete', sender);
    });

    this.model.onCompleting.add((sender) => {
      this.$emit('completing', sender);
    });

    if (this.mode === 'display') {
      this.model.mode = 'display';
    }

    const defaultThemeColors = SurveyVue.StylesManager.ThemeColors.default;
    defaultThemeColors['$border-color'] = theme.colors.gray['5'];
    defaultThemeColors['$body-container-background-color'] =
      theme.colors.gray['7'];
    defaultThemeColors['$disable-color'] = theme.colors.black;
    defaultThemeColors['$disabled-label-color'] = theme.colors.gray['3'];
    defaultThemeColors['$disabled-switch-color'] = 'rgba(72, 29, 115, 0.5)';
    defaultThemeColors['$disabled-slider-color'] = theme.colors.white;
    defaultThemeColors['$error-background-color'] = theme.colors['error-light'];
    defaultThemeColors['$error-color'] = theme.colors.error;
    defaultThemeColors['$inputs-background-color'] = theme.colors.white;
    defaultThemeColors['$main-color'] = theme.colors.primary;
    defaultThemeColors['$main-hover-color'] = theme.colors['primary-tint']['1'];
    defaultThemeColors['$slider-color'] = theme.colors.white;
    defaultThemeColors['$text-color'] = theme.colors.black;
    SurveyVue.StylesManager.applyTheme();

    this.model.css = {
      checkbox: {
        itemDecorator: 'sv-svg',
      },
    };

    // Other possible theme values to be declared:
    // $header-background-color: "#e7e7e7"
    // $header-color: "#6d7072"
    // $progress-text-color: "#9d9d9d"
  },
};
</script>

<style>
.base-survey.sv_main {
  @apply font-body-medium font-sans;
}

.base-survey.sv_main .sv_body {
  @apply border-none p-0;
}

.base-survey.sv_main input[type='button'],
.base-survey.sv_main button {
  @apply duration-ludicrous
    ease-out
    font-semibold
    px-4
    h-8
    rounded
    text-14
    transition-all;
}

.base-survey.sv_main
  .sv_body
  .sv_p_root
  .sv_q
  input:not([type='button']):not([type='reset']):not([type='submit']):not([type='image']):not([type='checkbox']):not([type='radio']):hover,
.base-survey.sv_main textarea:hover {
  @apply border-gray-3;
}

.base-survey.sv_main .sv-boolean__switch {
  @apply duration-light
    ease-out
    inline-flex
    items-center
    mx-2
    transition-colors
    w-12;
}

.base-survey.sv_main .sv-boolean--indeterminate .sv-boolean__switch {
  @apply bg-gray-4;
}

.base-survey.sv_main textarea:focus,
.base-survey.sv_main .sv-boolean input:focus ~ .sv-boolean__switch,
.base-survey.sv_main input[type='text']:focus,
.base-survey.sv_main input[type='button']:focus,
.base-survey.sv_main button:focus {
  @apply outline-none shadow-none border-primary;
}

.base-survey.sv_main .sv_container .sv_body .sv_p_root .sv_q textarea {
  @apply appearance-none font-body-text font-sans p-4;
}

.base-survey.sv_main
  .sv_q_erbox:not([style*='display: none']):not([style*='display:none']) {
  @apply bg-error-light border-none p-4 rounded-md text-error;
}

.base-survey.sv_main .sv-boolean__slider {
  @apply h-4 w-4;
}

.base-survey.sv_main .sv-boolean--indeterminate .sv-boolean__slider {
  margin-left: calc(50% - 8px);
}

.base-survey.sv_main .sv-boolean--checked .sv-boolean__slider {
  margin-left: calc(100% - 16px);
}

.base-survey.sv_main .sv_container .sv_body .sv_p_root .sv_q {
  @apply px-6 py-4;
}

.base-survey.sv_main input[type='radio'] {
  @apply sr-only;
}

.base-survey.sv_main .sv_q_radiogroup_label .circle {
  @apply align-middle
    border
    border-gray-4
    border-solid
    box-border
    h-6
    inline-flex
    items-center
    justify-center
    mr-4
    relative
    rounded-full
    w-6;
}

.base-survey.sv_main .sv_q_radiogroup_label .circle svg {
  @apply hidden;
}

.base-survey.sv_main .checked .sv_q_radiogroup_label .circle {
  @apply border-2 border-primary;
}

.base-survey.sv_main .checked .sv_q_radiogroup_label .circle::before {
  @apply absolute bg-primary block h-0 rounded-full w-1/2;
  content: '';
  padding-top: 50%;
}

.base-survey.sv_main .sv_body .sv_p_root .sv_q input[type='checkbox'] {
  @apply hidden;
}

.base-survey.sv_main .sv_body .sv_q_checkbox .sv_q_checkbox_label {
  @apply relative flex items-center;
}

.base-survey.sv_main .sv_body .sv_q_checkbox {
  @apply mb-2;
}

.base-survey.sv_main .sv_body .sv_q_checkbox:last-child {
  @apply mb-0 !important;
}

.base-survey.sv_main .sv_body .sv_q_checkbox .form-group {
  @apply mt-4;
}

.base-survey.sv_main
  .sv_body
  .sv_q_checkbox
  .sv_q_checkbox_label
  > span:last-child {
  @apply leading-6 pl-10;
}

.base-survey.sv_main
  .sv_body
  .sv_q_checkbox
  .sv_q_checkbox_label
  .checkbox-material {
  @apply absolute left-0 w-6 h-6 flex items-center justify-center bg-white text-white border border-solid border-gray-4 rounded-md;
}

.base-survey.sv_main
  .sv_body
  .sv_q_checkbox
  .sv_q_checkbox_label
  .checkbox-material
  .sv-svg {
  @apply w-6 h-6 text-white fill-current;
}

.base-survey.sv_main
  .sv_body
  .sv_q_checkbox.checked
  .sv_q_checkbox_label
  .checkbox-material {
  @apply bg-primary border-primary;
}
</style>
