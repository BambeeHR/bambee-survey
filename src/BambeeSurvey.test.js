import { mount } from '@vue/test-utils';
import BambeeSurvey from './BambeeSurvey.vue';
import survey from './assets/surveys/HrAudit.json';

describe('BambeeSurvey', () => {
  describe('component', () => {
    test('renders correctly', () => {
      const wrapper = mount(BambeeSurvey, { propsData: { survey } });
      expect(wrapper).toMatchSnapshot();
    });

    test('emits an input event when data changes', async () => {
      const wrapper = mount(BambeeSurvey, { propsData: { survey } });
      wrapper.vm.$data.model.setValue('hasRemoteWorkers', true);

      expect(wrapper.emitted().input[0][1]).toEqual(
        expect.objectContaining({
          name: 'hasRemoteWorkers',
          value: true,
        })
      );
    });
    // TODO(jon.jandoc): Need to add test coverage for the complete event.

    test.skip('advances to most recent page when passing initial values', () => {
      const wrapper = mount(BambeeSurvey, {
        propsData: {
          survey,
          initialValues: {
            hasPayrollSolution: false,
            hasPosters: 'no',
            hasPtoPolicy: 'yes',
            hasRemoteWorkers: false,
          },
        },
      });
      expect(
        wrapper.find('input[name="hasPayrollSolution"]').element
      ).toBeVisible();
    });
  });
});
