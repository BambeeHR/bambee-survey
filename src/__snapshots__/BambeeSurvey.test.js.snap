// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BambeeSurvey component advances to most recent page when passing initial values 1`] = `
<div class="base-survey sv_main sv_default_css">
  <form onsubmit="return false;">
    <div class="sv_custom_header"></div>
    <div class="sv_container">
      <!---->
      <!---->
      <div class="sv_body">
        <!---->
        <!---->
        <!---->
        <div class="sv_p_root">
          <!---->
          <!---->
          <div class="sv_row">
            <div class="sv-vue-row-additional-div">
              <div class="sv_q sv_qstn" id="sq_146" aria-labelledby="sq_146_ariaTitle" name="hasPayrollSolution" style="flex-basis: 100.000000%; flex-grow: 1; flex-shrink: 1; width: 100.000000%; min-width: 300px; max-width: initial; display: inline-block;">
                <div class="">
                  <h5 aria-label="Do you currently use a payroll solution? " id="sq_146_ariaTitle" class="sv_q_title">
                    <div>
                      <div>
                        <!----> <span aria-hidden="true" class="sv_q_num" style="position: static;">4.</span>
                        <!----> <span style="position: static;"><span style="position: static;">Do you currently use a payroll solution? </span></span> <span aria-hidden="true" class="sv_q_required_text">*</span>
                      </div> <span tabindex="0" aria-expanded="true" class="sv_question_icon sv_expanded" style="display: none;"></span>
                    </div>
                  </h5>
                  <div class="sv_q_description"><span style="position: static;"><span style="position: static;"></span></span></div>
                </div>
                <div class="">
                  <div role="alert" id="sq_146_errors" class="sv_q_erbox sv_qstn_error_top" style="display: none;"></div>
                  <div class="sv_qcbc sv_qbln"><label class="sv-boolean"><input type="checkbox" name="hasPayrollSolution" id="sq_146i" aria-required="true" aria-label="Do you currently use a payroll solution? " class="sv-visuallyhidden" value="false"> <span class="sv-boolean__label  "><span style="position: static;"><span style="position: static;">No</span></span></span>
                      <div class="sv-boolean__switch"><span class="sv-boolean__slider"></span></div> <span class="sv-boolean__label  sv-boolean__label--disabled"><span style="position: static;"><span style="position: static;">Yes</span></span></span>
                    </label></div>
                  <!---->
                  <!---->
                  <!---->
                </div>
                <!---->
              </div>
            </div>
          </div>
          <!---->
          <!---->
        </div>
        <!---->
        <div class="sv_progress">
          <div role="progressbar" aria-valuemin="0" aria-valuemax="100" class="sv_progress_bar" style="width: 37%;"><span class="undefined sv-hidden">Page 4 of 11</span></div> <span class="undefined undefined">Page 4 of 11</span>
        </div>
        <div class="sv_nav"><input type="button" value="Previous" class=" sv_prev_btn"> <input type="button" value="Next" class=" sv_next_btn">
          <!----> <input type="button" value="Complete" class=" sv_complete_btn">
        </div>
      </div>
      <!---->
      <!---->
      <!---->
      <!---->
    </div>
  </form>
</div>
`;

exports[`BambeeSurvey component renders correctly 1`] = `
<div class="base-survey sv_main sv_default_css">
  <form onsubmit="return false;">
    <div class="sv_custom_header"></div>
    <div class="sv_container">
      <!---->
      <!---->
      <div class="sv_body">
        <!---->
        <!---->
        <!---->
        <div class="sv_p_root">
          <!---->
          <!---->
          <div class="sv_row">
            <div class="sv-vue-row-additional-div">
              <div style="flex-basis: 100.000000%; flex-grow: 1; flex-shrink: 1; width: 100.000000%; min-width: 300px; max-width: initial;">
                <div class="sv_q sv_qstn" id="sq_100" aria-labelledby="sq_100_ariaTitle" name="hasRemoteWorkers">
                  <div class="">
                    <h5 aria-label="Do you have any remote workers?" id="sq_100_ariaTitle" class="sv_q_title">
                      <div>
                        <!---->
                        <div>
                          <!----> <span aria-hidden="true" class="sv_q_num" style="position: static;">1.</span>
                          <!----> <span style="position: static;"><span class="sv-string-viewer" style="position: static;">Do you have any remote workers?</span></span> <span aria-hidden="true" class="sv_q_required_text">*</span>
                        </div>
                      </div>
                      <!---->
                    </h5>
                    <div class="sv_q_description"><span style="position: static;"><span class="sv-string-viewer" style="position: static;"></span></span></div>
                  </div>
                  <div class="">
                    <div role="alert" aria-live="polite" id="sq_100_errors" class="sv_q_erbox sv_qstn_error_top" style="display: none;"></div>
                    <div class="sv_qcbc sv_qbln"><label class="sv-boolean sv-boolean--indeterminate"><input type="checkbox" name="hasRemoteWorkers" id="sq_100i" aria-required="true" aria-label="Do you have any remote workers?" class="sv-visuallyhidden" value=""> <span class="sv-boolean__label "><span style="position: static;"><span class="sv-string-viewer" style="position: static;">No</span></span></span>
                        <div class="sv-boolean__switch"><span class="sv-boolean__slider"></span></div> <span class="sv-boolean__label "><span style="position: static;"><span class="sv-string-viewer" style="position: static;">Yes</span></span></span>
                      </label></div>
                    <!---->
                    <!---->
                    <!---->
                  </div>
                  <!---->
                </div>
                <!---->
              </div>
            </div>
          </div>
          <!---->
        </div>
        <!---->
        <div class="sv_progress">
          <div role="progressbar" aria-valuemin="0" aria-valuemax="100" class="sv_progress_bar" style="width: 10%;"><span class="sv-hidden">Page 1 of 11</span></div> <span class="">Page 1 of 11</span>
        </div>
        <div class="sv_nav"><input type="button" value="Previous" class="sv_prev_btn"> <input type="button" value="Next" class="sv_next_btn">
          <!----> <input type="button" value="Complete" class="sv_complete_btn">
        </div>
      </div>
      <!---->
      <!---->
      <!---->
      <!---->
    </div>
  </form>
</div>
`;
